azure/services/availabilitysets/availabilitysets.go

```diff
@@ -18,134 +18,98 @@ package availabilitysets
-	if err != nil {
-		return errors.Wrap(err, "failed to get availability sets sku")
-	}
-
-	faultDomainCountStr, ok := asSku.GetCapability(resourceskus.MaximumPlatformFaultDomainCount)
-	if !ok {
-		return errors.Errorf("cannot find capability %s sku %s", resourceskus.MaximumPlatformFaultDomainCount, *asSku.Name)
-	}
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultAzureServiceReconcileTimeout)
+	defer cancel()
 
-	faultDomainCount, err := strconv.ParseUint(faultDomainCountStr, 10, 32)
-	if err != nil {
-		return errors.Wrap(err, "failed to determine max fault domain count")
+	setSpec := s.Scope.AvailabilitySetSpec()
+	var err error
+	if setSpec != nil {
+		_, err = s.CreateResource(ctx, setSpec, serviceName)
+	} else {
+		log.V(2).Info("skip creation when no availability set spec is found")
 	}
 
-	log.V(2).Info("creating availability set", "availability set", availabilitySetName)
-
-	asParams := compute.AvailabilitySet{
-		Sku: &compute.Sku{
-			Name: to.StringPtr(string(compute.AvailabilitySetSkuTypesAligned)),
-		},
-		AvailabilitySetProperties: &compute.AvailabilitySetProperties{
-			PlatformFaultDomainCount: to.Int32Ptr(int32(faultDomainCount)),
-		},
-		Tags: converters.TagsToMap(infrav1.Build(infrav1.BuildParams{
-			ClusterName: s.Scope.ClusterName(),
-			Lifecycle:   infrav1.ResourceLifecycleOwned,
-			Name:        to.StringPtr(availabilitySetName),
-			Role:        to.StringPtr(infrav1.CommonRole),
-			Additional:  s.Scope.AdditionalTags(),
-		})),
-		Location: to.StringPtr(s.Scope.Location()),
-	}
-
-	_, err = s.Client.CreateOrUpdate(ctx, s.Scope.ResourceGroup(), availabilitySetName, asParams)
-	if err != nil {
-		return errors.Wrapf(err, "failed to create availability set %s", availabilitySetName)
-	}
-
-	log.V(2).Info("successfully created availability set", "availability set", availabilitySetName)
-
-	return nil
+	s.Scope.UpdatePutStatus(infrav1.AvailabilitySetReadyCondition, serviceName, err)
+	return err
 }
 
 // Delete deletes availability sets.
 func (s *Service) Delete(ctx context.Context) error {
 	ctx, log, done := tele.StartSpanWithLogger(ctx, "availabilitysets.Service.Delete")
 	defer done()
 
-	availabilitySetName, ok := s.Scope.AvailabilitySet()
-	if !ok {
-		return nil
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultAzureServiceReconcileTimeout)
+	defer cancel()
+
+	setSpec := s.Scope.AvailabilitySetSpec()
+	var resultingErr error
+	if setSpec == nil {
+		log.V(2).Info("skip deletion when no availability set spec is found")
+	} else {
+		existingSet, err := s.Client.Get(ctx, setSpec)
+		if err != nil {
+			if !azure.ResourceNotFound(err) {
+				resultingErr = errors.Wrapf(err, "failed to get availability set %s in resource group %s", setSpec.ResourceName(), setSpec.ResourceGroupName())
+			}
+		} else {
+			availabilitySet, ok := existingSet.(compute.AvailabilitySet)
+			if !ok {
+				resultingErr = errors.Errorf("%T is not a compute.AvailabilitySet", existingSet)
```

Comment:
```suggestion
				return errors.Errorf("%T is not a compute.AvailabilitySet", existingSet)
```

Reply:
oh nevermind, I see you did this to call UpdateDeleteStatus even when there's an error 🤔